File: 215. Kth Largest Element in an Array.cpp
class Solution {

struct cmp{
    bool operator()(int a, int b){
        return a > b;
    }    
};

public:
    // sort, O(NlogN)time, no space, 51.77%, ok
    int findKthLargest(vector<int>& nums, int k) {
        if(nums.size() < k) return -1;
        sort(nums.begin(), nums.end(), [](int a, int b){return a > b;});
        return nums[k - 1];
    }
    
    // priority queue, O(NlogK)time, O(K) space, 72.12%, ok
    int findKthLargest(vector<int>& nums, int k) {
        if(nums.size() < k) return -1;
        priority_queue<int, vector<int>, cmp> pq;
        for(auto n: nums){
            pq.push(n);
            if(pq.size() > k) pq.pop();
        }
        return pq.top();
    }
    
    
};

File: 215. Kth Largest Element in an Array.py
import heapq

class Solution(object):
    
    # sort: O(NlogN) time, no space
    def findKthLargest(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        nums.sort()
        return nums[-k]
        
    # min heap: O(NlogK) time, O(k) space
    def findKthLargest(self, nums, k):
        ret = [float('-inf')] * k
        for num in nums:
            heapq.heappushpop(ret, num)
        return ret[0]
        